File: packages/create-wayforth-agent/templates/python/translate-notify/agent.py
"""
{{AGENT_NAME}} — translate-notify agent

Discovers the best translation service via Wayforth at runtime,
translates text through the proxy (native upstream shape + automatic failover),
and optionally sends a notification email.

Pattern: discovery via SDK search  ·  execution via /proxy/{slug}
"""
import os
import sys
import httpx
from wayforth import Wayforth

API_KEY = os.environ.get("WAYFORTH_API_KEY", "")
if not API_KEY:
    sys.exit("Set WAYFORTH_API_KEY in your environment. Get a key at wayforth.io/signup")

TEXT_TO_TRANSLATE = "Hello from Wayforth — your agent is live."
TARGET_LANG       = "FR"   # DeepL language code

# Optional: set to send a notification email via Resend
NOTIFY_EMAIL = os.environ.get("NOTIFY_EMAIL", "")

_PROXY_CATALOG_SLUGS = {
    "alphavantage", "assemblyai", "brave_search", "deepl", "elevenlabs", "firecrawl",
    "gemini", "groq", "jina", "mistral", "openweather", "perplexity",
    "resend", "serper", "stability", "tavily", "together",
}
_CATALOG_TO_PROXY = {"brave_search": "brave"}


def proxy_post(slug: str, params: dict) -> httpx.Response:
    resp = httpx.post(
        f"https://gateway.wayforth.io/proxy/{slug}",
        headers={"X-Wayforth-API-Key": API_KEY},
        json=params,
        timeout=30,
    )
    resp.raise_for_status()
    return resp


def main() -> None:
    wf = Wayforth(api_key=API_KEY)

    # ── 1. DISCOVER translation service ────────────────────────────────────
    print("Discovering best translation service...")
    results = wf.search("translate text to another language", limit=10)["results"]
    candidates = [r for r in results if r.get("slug") in _PROXY_CATALOG_SLUGS]
    if not candidates:
        sys.exit("No proxy-managed translation service found in discovery results.")

    # ── 2. TRANSLATE via proxy (try candidates in WRI order) ───────────────
    resp = None
    slug = None
    for hit in candidates:
        slug = hit["slug"]
        print(f"→ Trying {hit['name']}  WRI={hit['wri']}  slug={slug}")
        r = httpx.post(
            f"https://gateway.wayforth.io/proxy/{slug}",
            headers={"X-Wayforth-API-Key": API_KEY},
            json={"text": TEXT_TO_TRANSLATE, "target_lang": TARGET_LANG},
            timeout=30,
        )
        if r.is_success:
            resp = r
            break
        print(f"  ✗ {slug} returned {r.status_code} — trying next service")

    if resp is None:
        sys.exit("All discovered services failed. Check your API key and credits.")

    print(f"\nTranslating via /proxy/{slug}...")

    print(f"[wayforth] failover  : {resp.headers.get('x-wayforth-failover')}")
    print(f"[wayforth] wri       : {resp.headers.get('x-wayforth-wri')}")
    print(f"[wayforth] credits   : {resp.headers.get('x-wayforth-credits-remaining')} remaining")

    # Native DeepL adapter shape: {"translated_text": "...", "detected_source_lang": "..."}
    data            = resp.json()
    translated_text = data.get("translated_text", "")
    source_lang     = data.get("detected_source_lang", "?")

    print(f"\nOriginal  ({source_lang}): {TEXT_TO_TRANSLATE}")
    print(f"Translated ({TARGET_LANG}): {translated_text}")

    # ── 3. NOTIFY (optional) ───────────────────────────────────────────────
    if not NOTIFY_EMAIL:
        print("\n[skip] Set NOTIFY_EMAIL to also send a notification via Resend.")
        return

    print(f"\nDiscovering notification service...")
    notif_results = wf.search("send email notification", limit=10)["results"]
    notif_hit = next((r for r in notif_results if r.get("slug") in _PROXY_CATALOG_SLUGS), None)
    if not notif_hit:
        print("[skip] No proxy-managed email service found.")
        return

    notif_slug = notif_hit["slug"]
    print(f"→ {notif_hit['name']}  slug={notif_slug}")

    notif_resp = proxy_post(notif_slug, {
        "to":      NOTIFY_EMAIL,
        "subject": f"Translation complete: {TARGET_LANG}",
        "html":    f"<p><strong>Original:</strong> {TEXT_TO_TRANSLATE}</p>"
                   f"<p><strong>Translated:</strong> {translated_text}</p>",
    })
    print(f"Notification sent: {notif_resp.json()}")


if __name__ == "__main__":
    main()

File: packages/create-wayforth-agent/templates/typescript/translate-notify/src/agent.ts
/**
 * {{AGENT_NAME}} — translate-notify agent
 *
 * Discovers the best translation service via Wayforth at runtime,
 * translates text through the proxy (native upstream shape + failover),
 * and optionally sends a notification email.
 *
 * Pattern: discovery via SDK search · execution via /proxy/{slug}
 */
import { WayforthClient } from 'wayforth-sdk';

const API_KEY = process.env.WAYFORTH_API_KEY ?? '';
if (!API_KEY) {
  console.error('Set WAYFORTH_API_KEY in your environment. Get a key at wayforth.io/signup');
  process.exit(1);
}

const TEXT_TO_TRANSLATE = 'Hello from Wayforth — your agent is live.';
const TARGET_LANG       = 'FR';
const NOTIFY_EMAIL      = process.env.NOTIFY_EMAIL ?? '';
const GATEWAY           = 'https://gateway.wayforth.io';

const PROXY_CATALOG_SLUGS = new Set([
  'alphavantage', 'assemblyai', 'brave_search', 'deepl', 'elevenlabs', 'firecrawl',
  'gemini', 'groq', 'jina', 'mistral', 'openweather', 'perplexity',
  'resend', 'serper', 'stability', 'tavily', 'together',
]);
const CATALOG_TO_PROXY: Record<string, string> = { brave_search: 'brave' };

async function proxyPost(slug: string, params: Record<string, unknown>) {
  const resp = await fetch(`${GATEWAY}/proxy/${slug}`, {
    method: 'POST',
    headers: { 'X-Wayforth-API-Key': API_KEY, 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  });
  if (!resp.ok) throw new Error(`${slug} proxy returned ${resp.status}: ${await resp.text()}`);
  return resp;
}

async function main() {
  const wf = new WayforthClient({ apiKey: API_KEY });

  // ── 1. DISCOVER translation service ──────────────────────────────────────
  console.log('Discovering best translation service...');
  const { results } = await wf.search('translate text to another language', { limit: 10 });
  const candidates = results.filter((r) => r.slug !== null && PROXY_CATALOG_SLUGS.has(r.slug!));
  if (!candidates.length) throw new Error('No proxy-managed translation service found.');

  // ── 2. TRANSLATE via proxy (try candidates in WRI order) ─────────────────
  let resp: Response | null = null;
  let usedSlug = '';
  for (const hit of candidates) {
    const slug = hit.slug!;
    console.log(`→ Trying ${hit.name}  WRI=${hit.wri}  slug=${slug}`);
    const r = await fetch(`${GATEWAY}/proxy/${slug}`, {
      method: 'POST',
      headers: { 'X-Wayforth-API-Key': API_KEY, 'Content-Type': 'application/json' },
      body: JSON.stringify({ text: TEXT_TO_TRANSLATE, target_lang: TARGET_LANG }),
    });
    if (r.ok) { resp = r; usedSlug = slug; break; }
    console.log(`  ✗ ${slug} returned ${r.status} — trying next service`);
  }
  if (!resp) throw new Error('All discovered services failed. Check your API key and credits.');
  console.log(`\nTranslating via /proxy/${usedSlug}...`);
  console.log(`[wayforth] failover  : ${resp.headers.get('x-wayforth-failover')}`);
  console.log(`[wayforth] wri       : ${resp.headers.get('x-wayforth-wri')}`);
  console.log(`[wayforth] credits   : ${resp.headers.get('x-wayforth-credits-remaining')} remaining`);

  const data = await resp.json() as Record<string, string>;
  const translatedText = data['translated_text'] ?? '';
  const sourceLang     = data['detected_source_lang'] ?? '?';
  console.log(`\nOriginal  (${sourceLang}): ${TEXT_TO_TRANSLATE}`);
  console.log(`Translated (${TARGET_LANG}): ${translatedText}`);

  // ── 3. NOTIFY (optional) ─────────────────────────────────────────────────
  if (!NOTIFY_EMAIL) {
    console.log('\n[skip] Set NOTIFY_EMAIL to also send a notification via Resend.');
    return;
  }
  console.log('\nDiscovering notification service...');
  const { results: notifResults } = await wf.search('send email notification', { limit: 10 });
  const notifHit = notifResults.find((r) => r.slug !== null && PROXY_CATALOG_SLUGS.has(r.slug!));
  if (!notifHit?.slug) { console.log('[skip] No proxy-managed email service found.'); return; }

  const notifResp = await proxyPost(notifHit.slug, {
    to: NOTIFY_EMAIL, subject: `Translation complete: ${TARGET_LANG}`,
    html: `<p><strong>Original:</strong> ${TEXT_TO_TRANSLATE}</p><p><strong>Translated:</strong> ${translatedText}</p>`,
  });
  console.log('Notification sent:', await notifResp.json());
}

main().catch((e) => { console.error(e.message); process.exit(1); });
